Sources/AppcuesKit/Presentation/Extensions/UIScrollView+ScrollObserver.swift

```diff
@@ -0,0 +1,237 @@
+
+        if originalMethod == nil {
+            // this is the case where the existing delegate does not have an implementation for the target selector
+
+            guard let placeholderMethod = class_getInstanceMethod(UIScrollView.self, placeholderSelector) else {
+                // this really shouldn't ever be nil, as that would mean the function defined a few lines below is no
+                // longer there, but we must nil check this call
+                return
+            }
+
+            // add the placeholder, so it can be swizzled uniformly
+            class_addMethod(
+                type(of: delegate),
+                targetSelector,
+                method_getImplementation(placeholderMethod),
+                method_getTypeEncoding(placeholderMethod)
+            )
+        }
+
+        // swizzle the new implementation to inject our own custom logic
+
+        // this should never be nil, as it would mean the function defined a few lines below is no longer there,
+        // but we must nil check this call.
+        guard let swizzleMethod = class_getInstanceMethod(UIScrollView.self, swizzleSelector) else { return }
+
+        // add the swizzled version - this will only succeed once for this instance, if its already there, we've already
+        // swizzled, and we can exit early in the next guard
+        let addMethodResult = class_addMethod(
+            type(of: delegate),
+            swizzleSelector,
+            method_getImplementation(swizzleMethod),
+            method_getTypeEncoding(swizzleMethod)
+        )
+
+        guard addMethodResult,
+              let originalMethod = originalMethod ?? class_getInstanceMethod(type(of: delegate), targetSelector),
+              let swizzledMethod = class_getInstanceMethod(type(of: delegate), swizzleSelector) else {
+            return
+        }
+
+        // finally, here is where we swizzle in our custom implementation
+        method_exchangeImplementations(originalMethod, swizzledMethod)
+    }
+
+    @objc
+    func appcues__placeholderScrollViewDidScroll(_ scrollView: UIScrollView) {
+        // this gives swizzling something to replace, if the existing delegate doesn't already
+        // implement this function.
+    }
+
+    @objc
+    func appcues__placeholderScrollViewDidEndDecelerating(_ scrollView: UIScrollView) {
+        // this gives swizzling something to replace, if the existing delegate doesn't already
+        // implement this function.
+    }
+
+    @objc
+    func appcues__placeholderScrollViewDidEndDragging(_ scrollView: UIScrollView, willDecelerate decelerate: Bool) {
+        // this gives swizzling something to replace, if the existing delegate doesn't already
+        // implement this function.
+    }
+
+    @objc
+    func appcues__scrollViewDidScroll(_ scrollView: UIScrollView) {
+        appcues__scrollViewDidScroll(scrollView)
+
+        AppcuesScrollViewDelegate.shared.didScroll()
+    }
```

Comment:
Just tried it out and I believe it will work! nice suggestion. The case I was concerned with was the bounce effect after flinging to the end of a list. This appears to be handled though, as the `scrollViewDidEndDragging` is hit with `willDecelerate: true` - so we _do not_ send the call to `scrollEnded` at that point. It will then wait until `scrollViewDidEndDecelerating` as expected, and trigger the retry once the bounce has settled.